expertsystem/topology.py

```diff
@@ -22,23 +23,12 @@
 )
 
 
+@dataclass
 class Edge:
-    """Struct-like definition of an edge."""
+    """Struct-like definition of an edge, used in `Topology`."""
 
-    def __init__(self) -> None:
-        self.ending_node_id: Optional[int] = None
-        self.originating_node_id: Optional[int] = None
-
-    def __repr__(self) -> str:
-        return f"<class 'Edge': {self.ending_node_id} -> {self.originating_node_id}>"
-
-    def __eq__(self, other: object) -> bool:
-        if isinstance(other, Edge):
-            return (
-                self.ending_node_id == other.ending_node_id
-                and self.originating_node_id == other.originating_node_id
-            )
-        raise NotImplementedError
+    ending_node_id: Optional[int] = None
+    originating_node_id: Optional[int] = None
```

Comment:
Ideal would be to avoid having the state (None, None), which is def invalid

Reply:
Currently not possible because of the way in which `Topology` is built up elsewhere in the system through `add_edge`:

https://github.com/ComPWA/expertsystem/blob/5bca7a79be63fe5f18b2c10e151dce66b3f6dd62/expertsystem/topology.py#L91

Reply:
Ok, lets put that on the Todolist. Lets open a new Epic issue for that

Reply:
Should that really be an epic?

Reply:
those two classes are still quite unsafe

Reply:
Yes that's true. I'm currently working on the `Topology.verify` method, to address at least some of those issues.

Reply:
We can also make the way of building an invalid graph impossible, by changing the interface on how to add edges and nodes etc. For example these rules are quite solid:
- an Edge has to be attached to at least one node
- a new Node has to be attached to at least one edge (except the first node/seed node)

Reply:
Discussion moved to #250